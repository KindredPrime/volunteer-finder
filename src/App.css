@import url('https://fonts.googleapis.com/css2?family=Karla&family=Sura&family=Varela+Round&display=swap');

* {
  box-sizing: border-box;
}

body {
  background-color: #1A4161;
  color: white;
  margin-bottom: 20px;
}

h1 {
  text-shadow: -3px 0 black, 0 3px black, 3px 0 black, 0 -3px black;
  font-size: 34px;
  line-height: 48px;
}

h2 {
  text-shadow: -2px 0 black, 0 2px black, 2px 0 black, 0 -2px black;
  font-size: 26px;
  line-height: 36px;
}

p, legend {
  text-shadow: -1px 0 black, 0 1px black, 1px 0 black, 0 -1px black;
}

h1, h2, legend {
  font-family: 'Varela Round', sans-serif;
}

legend {
  font-size: 18px;
  line-height: 24px;
}

p, label {
  font-family: 'Karla', sans-serif;
}

p, label {
  font-size: 16px;
  line-height: 24px;
}

p {
  text-align: left;
}

section {
  border-top: 3px solid #747A82;
  margin: 20px 0px 20px 0px;
}

section > * {
  margin-left: 5%;
  margin-right: 5%;
}

button {
  background-color: #FF9100;
  color: black;
  cursor: pointer;
}

button:disabled {
  background-color: #747A82;
}

a {
  text-decoration: none;
  color: white;
}

a, button {
  font-family: 'Sura', serif;
}

form {
  padding: 10px;
}

fieldset, legend {
  margin-left: auto;
  margin-right: auto;
}

label {
  margin-right: 10px;
}

form button {
  margin-top: 10px;
}

fieldset {
  border: none;
}

legend {
  font-weight: bold;
}

p.error {
  color: #FF9100;
}

ul {
  padding: 0;
}

li {
  list-style: none;
}

.App {
  margin-top: 75px;
  text-align: center;
}

.App.expanded {
  margin-top: 138px;
}

@media screen and (min-width: 750px) {
  .App.expanded {
    margin-top: 75px;
  }
}

@media screen and (min-width: 1200px) {
  .App, .App.expanded {
    margin-top: 0px;
    margin-left: 17%;
  }
}