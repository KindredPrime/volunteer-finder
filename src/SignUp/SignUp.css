form {
  background-color: #D7D7D7;
}

.SignUp__error {
  color: red;
}