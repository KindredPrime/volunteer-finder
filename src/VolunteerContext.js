import React from 'react';

const VolunteerContext = React.createContext({
  orgs: [],
  causes: [],
  addOrg: () => {},
  deleteOrg: () => {},
  appError: null,
  // Is the app fetching data from the API?
  fetching: false
});

export default VolunteerContext;