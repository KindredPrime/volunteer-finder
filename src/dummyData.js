const dummyUsers = [
  {
    id: 1,
    email: 'email1@email.com',
    username: 'user1',
    password: 'password1',
    orgsAdded: [1],
    eventsAdded: [1, 2]
  },
  {
    id: 2,
    email: 'email2@email.com',
    username: 'user2',
    password: 'password2',
    orgsAdded: [2],
    eventsAdded: [3]
  },
  {
    id: 3,
    email: 'email3@email.com',
    username: 'user3',
    password: 'password3',
    orgsAdded: [3],
    eventsAdded: []
  },
  {
    id: 4,
    email: 'email4@email.com',
    username: 'user4',
    password: 'password4',
    orgsAdded: [],
    eventsAdded: []
  },
  {
    id: 5,
    email: 'email5@email.com',
    username: 'user5',
    password: 'password5',
    orgsAdded: [4],
    eventsAdded: [4]
  }
];

const dummyOrgs = [
  {
    id: 1,
    name: 'YMCA',
    website: 'https://www.ymca.com',
    phone: '1-800-555-7777',
    email: 'contact@ymca.com',
    address: '1 YMCA Street Real City, Real State',
    causes: [1, 2, 3],
    tags: [2, 3, 4, 6]
  },
  {
    id: 2,
    name: 'ASPCA',
    website: 'https://www.aspca.com',
    phone: '1-800-123-4567',
    email: 'contact@aspca.com',
    address: '1 ASPCA Street Real City, Real State',
    causes: [5, 2],
    tags: [2, 3, 4, 10]
  },
  {
    id: 3,
    name: 'National VOAD',
    website: 'https://www.nvoad.com',
    phone: '1-800-456-7890',
    email: 'contact@nvoad.com',
    address: '1 NVOAD Street Rivendell, Middle Earth',
    causes: [4, 2],
    tags: [2, 3, 4, 5, 7, 11]
  },
  {
    id: 4,
    name: 'National Air and Space Museum',
    website: 'https://www.airandspace.com',
    phone: '1-800-987-6543',
    email: 'contact@airandspace.com',
    address: '1 Space Street Air City, Outer Space',
    causes: [6, 7],
    tags: [1, 2, 3, 4, 6, 12]
  }
];

const dummyEvents = [
  {
    id: 1,
    name: 'Member Appreciation Day',
    organization: 1,
    location: 'Reston YMCA Center',
    date: 'February 20, 2020 00:00:00',
    duration: '4 hours',
    description: 'Greet YMCA members as they enter the facility and offer them snacks.  See https://www.website.com for more details.',
    causes: [2, 3],
    tags: [2, 4, 12]
  },
  {
    id: 2,
    name: 'Triathlon Assistance',
    organization: 1,
    location: 'Reston YMCA Center',
    date: 'January 27, 2020 00:00:00',
    duration: '4 hours',
    description: 'Help YMCA staff host their annual indoor triathlon: sign in triathlon participants, direct the participants to the different stages of the triathlon, and record times. See https://www.triathlon.com for more details.',
    causes: [2, 3],
    tags: [2, 3, 9]
  },
  {
    id: 3,
    name: 'Adoption Day',
    organization: 2,
    location: 'Times Square, New York City, New York',
    date: 'December 13, 2020 00:00:00',
    duration: '8 hours',
    description: 'Help the ASPCA run their annual adoption day for their local pet shelter.',
    causes: [5],
    tags: [2, 3, 9, 12]
  },
  {
    id: 4,
    name: 'Docent Training Class',
    organization: 4,
    location: '14390 Air and Space Museum Pkwy Chantilly, VA 20151',
    date: 'January 30, 2021 00:00:00',
    duration: '2 hours',
    description: 'This is a required class for anyone who wants to be a docent (tour guide) at the National Air and Space Museum.',
    causes: [6, 7],
    tags: [2, 3, 13]
  }
];

const dummyTags = [
  {
    id: 1,
    name: 'virtual'
  },
  {
    id: 2,
    name: 'in-person'
  },
  {
    id: 3,
    name: 'weekends'
  },
  {
    id: 4,
    name: 'weekdays'
  },
  {
    id: 5,
    name: 'nights'
  },
  {
    id: 6,
    name: 'group'
  },
  {
    id: 7,
    name: 'delivery'
  },
  {
    id: 8,
    name: 'cooking'
  },
  {
    id: 9,
    name: 'administration'
  },
  {
    id: 10,
    name: 'cleaning'
  },
  {
    id: 11,
    name: 'heavy lifting'
  },
  {
    id: 12,
    name: 'greeting'
  },
  {
    id: 13,
    name: 'training'
  }
];

const dummyCauses = [
  {
    id: 1,
    name: 'Youth'
  },
  {
    id: 2,
    name: 'Community'
  },
  {
    id: 3,
    name: 'Health'
  },
  {
    id: 4,
    name: 'Disaster Response'
  },
  {
    id: 5,
    name: 'Animals'
  },
  {
    id: 6,
    name: 'Museums'
  },
  {
    id: 7,
    name: 'Education'
  },
  {
    id: 8,
    name: 'Human Rights'
  },
  {
    id: 9,
    name: 'Arts'
  }
];

export {
  dummyUsers,
  dummyOrgs,
  dummyEvents,
  dummyTags,
  dummyCauses
};