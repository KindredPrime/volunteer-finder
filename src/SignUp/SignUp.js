// This component is being excluded from the client until the API is built
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import VolunteerContext from '../VolunteerContext';
import Nav from '../Nav/Nav';
import ValidationError from '../ValidationError/ValidationError';
import './SignUp.css';

class SignUp extends Component {
  static contextType = VolunteerContext;

  state = {
    userEmail: {
      value: '',
      touched: false
    },
    username: {
      value: '',
      touched: false
    },
    password: {
      value: '',
      touched: false
    },
    confirmPassword: {
      value: '',
      touched: false
    },
    error: null
  };

  updateField = (fieldName, fieldValue) => {
    this.setState({
      [fieldName]: {
        value: fieldValue,
        touched: true
      }
    });
  };

  validateRequiredField = (fieldName) => {
    const field = this.state[fieldName];
    const fieldValue = field.value.trim();
    if (fieldValue.length === 0) {
      return `${_.capitalize(fieldName)} is required`;
    }
  };

  validateConfirmPassword = () => {
    const { password, confirmPassword } = this.state;
    if (confirmPassword.value !== password.value) {
      return `Passwords don't match`;
    }
  };

  handleSubmit = (e) => {
    e.preventDefault();

    // Implement this after the API is built
    /*
    const { userEmail, username, password } = this.state;
    const { signUpUser } = this.context;
    
    signUpUser(userEmail.value, username.value, password.value);
    this.props.history.push(`/account`);
    */
  };

  render() {
    const { confirmPassword, error } = this.state;

    return (
      <div className="SignUp">
        <Nav />
        <header>
          <h1>Sign Up</h1>
        </header>

        <main>
          {error && <p className="SignUp__error">{error.message}</p>}
          <form onSubmit={this.handleSubmit}>
            <div>
              <label htmlFor="user-email">Email</label>
              <input
                type="email"
                id="user-email"
                onChange={(e) => this.updateField("userEmail", e.target.value, this)}
                required
              />
            </div>

            <br />


            <div>
              <label htmlFor="username">Username</label>
              <input
                type="text"
                id="username"
                onChange={(e) => this.updateField("username", e.target.value, this)}
                required
              />
            </div>

            <br />

            <div>
              <label htmlFor="password">Password</label>
              <input
                type="password"
                id="password"
                onChange={(e) => this.updateField("password", e.target.value, this)}
                required
              />
            </div>

            <br />

            <div>
              <label htmlFor="confirm-password">Confirm Password</label>
              <input
                type="password"
                id="confirm-password"
                onChange={(e) => this.updateField("confirmPassword", e.target.value, this)}
                required
              />
            </div>
            {confirmPassword.touched && <ValidationError message={this.validateConfirmPassword()} />}

            <button 
              type="submit"
              disabled={
                this.validateRequiredField('userEmail')
                || this.validateRequiredField('username')
                || this.validateRequiredField('password')
                || this.validateConfirmPassword()
              }
            >
              Sign Up
            </button>
          </form>
        </main>
      </div>
    );
  }
}

SignUp.propTypes = {
  history: PropTypes.shape({
    push: PropTypes.func
  }).isRequired
};

export default SignUp;
