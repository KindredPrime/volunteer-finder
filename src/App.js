import React, { Component } from 'react';
import { Route, Redirect } from 'react-router-dom';
import { dummyUsers, dummyOrgs, dummyCauses, dummyTags } from './dummyData';
import VolunteerContext from './VolunteerContext';
import Home from './Home/Home';
import Login from './Login/Login';
import SignUp from './SignUp/SignUp';
import Account from './Account/Account';
import Organization from './Organization/Organization';
import PublicUser from './PublicUser/PublicUser';
import OrgSearch from './OrgSearch/OrgSearch';
import AddOrg from './AddOrg/AddOrg';
import './App.css';

class App extends Component {
  static contextType = VolunteerContext;

  state = {
    user: null,
    users: dummyUsers,
    orgs: dummyOrgs,
    causes: dummyCauses,
    tags: dummyTags
  };

  loginUser = (username, password) => {
    const { users } = this.state;

    const user = users.find((elem) => (
      elem.username === username && elem.password === password
    ));

    if (user) {
      window.localStorage.setItem('userId', user.id);
      this.setState({
        user
      });
    }
    else {
      return 'Invalid username and password combination';
    }
  }

  logoutUser = () => {
    window.localStorage.removeItem('userId');
    this.setState({
      user: null
    });
  }

  signUpUser = (email, username, password) => {
    const { users } = this.state;

    const id = users.length + 1;
    users.push({
      id,
      email,
      username,
      password,
      orgsAdded: []
    });

    this.setState({
      users
    });

    this.loginUser(username, password);

    return id;
  }

  addOrg = (name, website, phone, email, address, causes, tags) => {
    const { user, users, orgs } = this.state;
    const id = orgs.length + 1;
    const org = {
      id,
      name,
      website: website || 'Not Available',
      phone: phone || 'Not Available',
      email: email || 'Not Available',
      address: address || 'Not Available',
      causes,
      tags
    };

    const newUser = {
      ...user,
      orgsAdded: [...user.orgsAdded, org.name]
    };
    const newUsers = users.map((elem) => {
      if (elem.id === user.id) {
        return newUser
      }

      return elem;
    });
    this.setState({
      user: newUser,
      users: newUsers,
      orgs: [...orgs, org]
    });

    return id;
  }

  componentDidMount() {
    const storedId = window.localStorage.getItem('userId');
    if (storedId) {
      const { users } = this.state;
      const user = users.find((elem) => elem.id === parseInt(storedId));
      this.setState({
        user
      });
    }
  }

  render() {
    const { user, users, orgs, causes, tags } = this.state;
    const contextValue = {
      user,
      users,
      orgs,
      causes,
      tags,
      loginUser: this.loginUser,
      logoutUser: this.logoutUser,
      signUpUser: this.signUpUser,
      addOrg: this.addOrg
    };

    return (
      <VolunteerContext.Provider value={contextValue}>
        <div className="App">
          <Route exact path="/" component={Home} />
          <Route path="/login" render={({ history }) => (
            !user ? <Login history={history} /> : <Redirect to="/account" />
          )} />
          <Route path="/signup" component={SignUp} />
          <Route path="/account" render={() => user ? <Account /> : <Redirect to="/login" />} />
          <Route path="/org/:id" component={Organization} />
          <Route path="/user/:id" component={PublicUser} />
          <Route path="/org-search" render={() => <OrgSearch pageLimit={10} />} />
          <Route path="/add-org" render={({ history }) => (
            window.localStorage.getItem('userId') ? <AddOrg history={history}/> : <Redirect to="/login" />
          )} />
        </div>
      </VolunteerContext.Provider>
    );
  }
}

export default App;
