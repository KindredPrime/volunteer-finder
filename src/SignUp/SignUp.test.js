import ReactDOM from 'react-dom';
import { BrowserRouter, MemoryRouter } from 'react-router-dom';
import { render, screen } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import userEvent from '@testing-library/user-event';
import SignUp from './SignUp';
import App from '../App';
import { dummyUsers, dummyOrgs } from '../dummyData';
import VolunteerContext from '../VolunteerContext';

describe('SignUp Component', () => {
  const testUsers = [
    {
      id: 1,
      email: 'unique-email@email.com',
      username: 'uniqueUsername',
      password: 'genericPassword'
    }
  ];

  const contextValue = {
    users: testUsers
  };

  let history;
  beforeEach(() => {
    history = createMemoryHistory();
  })

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <BrowserRouter>
        <SignUp history={history} />
      </BrowserRouter>,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  // Finish this test after the API is created
  it.skip('creates a new user and redirects to its page', () => {
    const contextValue = {
      users: dummyUsers,
      orgs: dummyOrgs
    };
    render(
      <VolunteerContext.Provider value={contextValue}>
        <MemoryRouter initialEntries={['/signup']}>
          <App />
        </MemoryRouter>
      </VolunteerContext.Provider>
    );

    userEvent.type(screen.getByLabelText('Email'), 'email@email.com');
    userEvent.type(screen.getByLabelText('Username'), 'username');
    userEvent.type(screen.getByLabelText('Password'), 'password');
    userEvent.type(screen.getByLabelText('Confirm Password'), 'password');
    userEvent.click(screen.getByRole('button', { name: 'Sign Up' }));

    expect(document.body).toMatchSnapshot();
  });

  it(`displays error when passwords don't match`, () => {
    render(
    <BrowserRouter>
      <SignUp history={history} />
    </BrowserRouter>);

    userEvent.type(screen.getByLabelText('Email'), 'email@email.com');
    userEvent.type(screen.getByLabelText('Username'), 'username');
    userEvent.type(screen.getByLabelText('Password'), 'password');
    userEvent.type(screen.getByLabelText('Confirm Password'), 'differentPassword');

    expect(screen.getByText(`Passwords don't match`)).toBeInTheDocument();
  });

  // Finish this test after the API is created
  it.skip('displays an error when the email is taken', () => {
    render(
      <VolunteerContext.Provider value={contextValue}>
        <BrowserRouter>
          <SignUp history={history} />
        </BrowserRouter>
      </VolunteerContext.Provider>
    );

    userEvent.type(screen.getByLabelText('Email'), 'unique-email@email.com');
    userEvent.type(screen.getByLabelText('Username'), 'newUsername');
    userEvent.type(screen.getByLabelText('Password'), 'password');
    userEvent.type(screen.getByLabelText('Confirm Password'), 'password');
    userEvent.click(screen.getByRole('button'));

    expect(screen.getByText('A user with that email already exists')).toBeInTheDocument();
  });

  // Finish this test after the API is created
  it.skip('displays an error when the username is taken', () => {
    render(
      <VolunteerContext.Provider value={contextValue}>
        <BrowserRouter>
          <SignUp history={history} />
        </BrowserRouter>
      </VolunteerContext.Provider>
    );

    userEvent.type(screen.getByLabelText('Email'), 'new-email@email.com');
    userEvent.type(screen.getByLabelText('Username'), 'uniqueUsername');
    userEvent.type(screen.getByLabelText('Password'), 'password');
    userEvent.type(screen.getByLabelText('Confirm Password'), 'password');
    userEvent.click(screen.getByRole('button'));

    expect(screen.getByText('That username is already taken')).toBeInTheDocument();
  });
});