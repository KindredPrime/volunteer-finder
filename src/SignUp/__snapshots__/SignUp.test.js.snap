// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp Component creates a new user and redirects to its page 1`] = `
<body>
  <div>
    <div
      class="App"
    >
      <div
        class="User"
      >
        <nav
          class="Nav"
        >
          <a
            href="/"
          >
            Home
          </a>
          <a
            aria-current="page"
            class="active"
            href="/user"
          >
            Account
          </a>
          <a
            href="/"
          >
            Logout
          </a>
        </nav>
        <header>
          <h1>
            Account
          </h1>
        </header>
        <section>
          <header>
            <h2>
              Account Info
            </h2>
          </header>
          <p>
            Username: 
            username
          </p>
          <p>
            Email: 
            email@email.com
          </p>
        </section>
        <section>
          <header>
            <h2>
              Organizations Added
            </h2>
            <ul
              class="User__orgs"
            />
          </header>
        </section>
        <section>
          <header>
            <h2>
              Events Added
            </h2>
            <ul
              class="User__events"
            />
          </header>
        </section>
      </div>
    </div>
  </div>
</body>
`;
